v8 spec changes. (#77)
algorandfoundation/specs: dev/ledger.md

minimum balance requirement is decreased equivalently when an account
closes out or deletes an app.

When opting in to an application, there is a base minimum balance increase
of 100,000 microalgos. There is an additional minimum balance increase based on
the `LocalStateSchema` for that application, described by following formula:

`28500 * schema.NumUint + 50000 * schema.NumByteSlice` microalgos.

When creating an application, there is a base minimum balance increase
of 100,000 microalgos. There is an additional minimum balance increase
of `100000 * ExtraProgramPages` microalgos.  Finally, there is an
additional minimum balance increase based on the `GlobalStateSchema`
for that application, described by the following formula:

`28500 * schema.NumUint + 50000 * schema.NumByteSlice` microalgos.

### Boxes

The box store is an associative array mapping keys of type (uint64 x
byte-array) to values of type byte-array. The key is a pair in which
the first value corresponds to an existing (or previously existing)
application ID, and the second is a _box name_, 1 to 64 bytes in
length.  The value is a byte-array of length not greater than 32,768.

> 

Reviewer: ```suggestion

Contrary to global/local state keys, empty box names are only used to increase the I/O budget (see below) and are not a valid name for a box.

```

Author: Added

> When an application executes an opcode that creates or destroys a box,

Reviewer: When does this min balance change takes place?
Is it immediate or only at the end of the transaction?
In particular, suppose I create a temporary 32k-box that I then destroy (in the same app call) and I don't have the min balance to pay for that. Would this fail or would this succeed?

Is the `global MinBalance` being updated after each opcode (this would only apply in the weird case where the application account has been rekeyed to a normal account since re-entrancy is forbidden)

Author: All min balance checks are done at the very end of a transaction _group_. However, `min_balance` and `acct_params_get MinBalance` report the updated value immediately.

Write tracking includes box creates, though. So if you make big boxes, you might be stopped by your write budget.

If you aren't, and you create and then delete, we don't have to hit the disk (and we don't).

- Asset IDs that the executing program may read asset parameters from. This
  field is encoded as msgpack field `apas`. The maximum number of entries in
  this field is 8.
- Box references that the executing program or any other program in
  the same group may access for reading or modification when the
  reference matches the running programs app ID. This field is encoded
  as msgpack field `apbx`, each element of which is encoded as a

>   msgpack object containing an application ID (`i`) and name

Reviewer 2: I may just be confused here but isn't the `i` actually the index into the `ForeignApps` array rather than the appID? I'm basing this off of the `transactions.BoxRef` object.

Author: You're right! I forgot my own encoding for a while.  I think it's correct elsewhere. But anyway, I'll clean it up, thanks.

          `ApplicationCall` transaction. The new programs are not executed in
          this transaction.  **SUCCEED.**

### Application Stateful Execution Semantics

- Before the execution of the first ApplicationCall transaction in a

>   group, the combined size all boxes referred to in the box references

Reviewer: ```suggestion
  group, the combined size of all boxes referred to in the box references
```

>   of all transactions in the group must be less than 1,024 times the

Reviewer: ```suggestion
  of all transactions in the group must be less than the I/O budget, i.e., 1,024 times the
```

  total number of box references in the group, or else the group
  fails.
- During the execution of an `ApprovalProgram` or `ClearStateProgram`,
  the application’s `LocalStateSchema` and `GlobalStateSchema` may
  never be violated. The program's execution will fail on the first
  instruction that would cause the relevant schema to be
  violated. Writing a `Bytes` value to a local or global [Key/Value
  Store][Key/Value Stores] such that the sum of the lengths of the key
  and value in bytes exceeds 128, or writing any value to a key longer
  than 64 bytes, will likewise cause the program to fail on the
  offending instruction.
- During the execution of an `ApprovalProgram`, the total size of all

>   pending "dirty" boxes (caused by creation or modification of a box)

Reviewer: This notion of dirty is introduced here.

```suggestion
 boxes that are created or modified
```

>   in the group must not exceed 1,024 times the total number of box

Reviewer: ```suggestion
  in the group must not exceed the I/O budget
```

>   references in the group, or else the group fails.  The program's

Reviewer: ```suggestion
  or else the group fails.  The program's
```

  execution will fail on the first instruction that would cause the
  constraint to be violated.
- Global state may only be read for the application ID whose program

>   is executing, or for an _available) application ID. An attempt to

Reviewer: ```suggestion
  is executing, or for an _available_ application ID. An attempt to
```

  read global state for another application that is not _available_
  will cause the program execution to fail.

> - Asset parameters may only be read for assets whose ID is

Author: Added.
